src/mapping/Partition.hpp

```diff
@@ -0,0 +1,315 @@
+  void setMappingFinished();
+
+  /// Number of input vertices this partition operates on
+  unsigned int getNumberOfInputVertices() const;
+
+  /// The center coordinate of this partition
+  std::array<double, 3> getCenterCoords() const;
+
+  /// Indicates whether there are valid output points for this partition
+  bool isEmpty() const;
+
+private:
+  precice::logging::Logger _log{"mapping::Partition"};
+
+  bool _hasComputedMapping = false;
+  bool _hasMappedData      = false;
+  bool _emptyPartition     = false;
+
+  /// Partition center
+  mesh::Vertex _center;
+
+  /// Partition radius
+  const double _radius;
+
+  /// Array with one entry per data component. A plain array would require knowledge about the number of components
+  std::array<Eigen::VectorXd, 3> resultVector;
+
+  /// vector containing the normalized weights (only for conservative mappings)
+  Eigen::VectorXd normalizedWeights;
+
+  // Stores the global IDs of the vertices so that we can apply a binary
+  // search in order to query specific objects
+  boost::container::flat_set<VertexID> inputIDs;
+  boost::container::flat_set<VertexID> outputIDs;
+
+  /// The RBF solver
+  RadialBasisFctSolver<RADIAL_BASIS_FUNCTION_T> _rbfSolver;
+
+  /// Polynomial treatment in the RBF solver
+  Polynomial _polynomial;
+
+  //TODO: Can probably be removed
+  RADIAL_BASIS_FUNCTION_T _basisFunction;
+
+  /// The weighting function
+  CompactPolynomialC2 _weightingFunction;
+};
+
+// --------------------------------------------------- HEADER IMPLEMENTATIONS
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+Partition<RADIAL_BASIS_FUNCTION_T>::Partition(
+    int               dimension,
+    mesh::Vertex      center,
+    double            radius,
+    double            parameter,
+    std::vector<bool> deadAxis,
+    Polynomial        polynomial,
+    unsigned int      targetSize,
+    mesh::PtrMesh     inputMesh,
+    mesh::PtrMesh     outputMesh)
+    : _center(center), _radius(radius), _polynomial(polynomial), _basisFunction(parameter), _weightingFunction(radius)
+{
+  // Disable integrated polynomial, as it might cause locally singular matrices
+  PRECICE_ASSERT(_polynomial != Polynomial::ON, "Integrated polynomial is not supported for PoU.")
+  PRECICE_ASSERT(deadAxis.size() == dimension);
+  PRECICE_DEBUG("Center coordinates: {}", _center.getCoords());
+  PRECICE_DEBUG("Partition radius: {}", _radius);
+
+  // Get vertices to be mapped
+  auto outIDs = outputMesh->index().getVerticesInsideBox(center, radius);
+  // Constructing the partition when we don't have evaluation points is pointless
+  auto inIDs = inputMesh->index().getVerticesInsideBox(center, radius);
+
+  // If we have only very few vertices (10% of the target), we consider this partition as empty
+  // TODO: The second empty criterion is potentially not suitable for conservative mappings
+  _emptyPartition = outIDs.size() == 0 || inIDs.size() < 0.1 * targetSize;
```

Comment:
To my understanding, the Partition doesn't is basically non-functional if `_emptyParition == true`.

Suggestion:

```cpp
std::optional<mapping::Partition> makePartition(...) {
  //...
  if (emptyParition) {
    return {};  
  }
  return Partition{...};
};
```

Then you only keep the Partitions that do anything in the PoU mapping.

Reply:
In principle we do this already right now.

```cpp
Partition p(...)
    if (!p.empty()) {
      _clusters.emplace_back(std::move(p));
    }
```
With the `Partition` constructor bailing out in case the partition is empty. Your suggestion would require a free function, right? The problem is, the `empty()` information is generated in the `Partition` constructor and using `std::optional` would still require a similar logic as above, where we first check if the object was constructed and afterwards move the valid object or skip it, correct?

Reply:
> Your suggestion would require a free function, right?

Correct.

> The problem is, the empty() information is generated in the Partition constructor and using std::optional would still require a similar logic as above, where we first check if the object was constructed and afterwards move the valid object or skip it, correct?

I see, but then, this information generation step could be moved to `makePartition` (or even a completely separate function) and the result passed to the constructor of the Partition.
Splitting this will make things easier to test too.

If this is not easily doable, then let's leave it as is.